runtime-sdk/modules/evm/src/state.rs

```
@@ -0,0 +1,121 @@
+pub const CONFIDENTIAL_STORAGES: &[u8] = &[0x04];
+
+#[cfg(feature = "confidential")]
+/// Confidential store key pair ID domain separation context base.
+pub const CONFIDENTIAL_STORE_KEY_PAIR_ID_CONTEXT_BASE: &[u8] = b"oasis-runtime-sdk/evm: state";
+#[cfg(feature = "confidential")]
+const CONTEXT_KEY_CONFIDENTIAL_STORE_INSTANCE_COUNT: &str = "evm.ConfidentialStoreCounter";
+
+/// The number of hash blocks that can be obtained from the current blockchain.
+pub const BLOCK_HASH_WINDOW_SIZE: u64 = 256;
+
+/// Get a typed store for the given address' storage.
+pub fn storage<'a, C: Context>(
+    ctx: &'a mut C,
+    address: &'a H160,
+) -> storage::TypedStore<impl storage::Store + 'a> {
+    #[cfg(feature = "confidential")]
+    {
+        confidential_storage(ctx, address)
+    }
+    #[cfg(not(feature = "confidential"))]
+    {
+        public_storage(ctx, address)
+    }
+}
+
+pub fn public_storage<'a, C: Context>(
+    ctx: &'a mut C,
+    address: &'a H160,
+) -> storage::TypedStore<impl storage::Store + 'a> {
+    storage::TypedStore::new(storage::HashedStore::<_, blake3::Hasher>::new(
+        contract_storage(ctx.runtime_state(), STORAGES, address),
+    ))
+}
+
+#[cfg(feature = "confidential")]
+pub fn confidential_storage<'a, C: Context>(
+    ctx: &'a mut C,
+    address: &'a H160,
+) -> storage::TypedStore<Box<dyn storage::Store + 'a>> {
+    fn empty_store() -> storage::TypedStore<Box<dyn storage::Store>> {
+        storage::TypedStore::new(Box::new(storage::EmptyStore::new()))
+    }
+    let kmgr_client = match ctx.key_manager() {
+        Some(kmgr_client) => kmgr_client,
+        None => return empty_store(),
+    };
+    let key_id = oasis_runtime_sdk::keymanager::get_key_pair_id(&[
+        CONFIDENTIAL_STORE_KEY_PAIR_ID_CONTEXT_BASE,
+        address.as_ref(),
+    ]);
+    let keypair = match kmgr_client.get_or_create_keys(key_id) {
+        Ok(keypair) => keypair,
+        Err(_) => return empty_store(),
```

This is actually a bad idea, it should return an aborting error instead (or panic). Otherwise this could make the node cause a discrepancy in case there is an error with the key manager (instead of indicating failure).